heron/schedulers/src/java/com/twitter/heron/scheduler/local/LocalScheduler.java

```diff
@@ -261,12 +262,15 @@ public boolean onUpdate(Scheduler.UpdateTopologyRequest request) {
   @Override
   public void addContainers(Set<PackingPlan.ContainerPlan> containers) {
     synchronized (processToContainer) {
+      Set<Integer> containerIds = new HashSet<>();
+      containerIds.addAll(processToContainer.values());
       for (PackingPlan.ContainerPlan container : containers) {
-        if (processToContainer.containsKey(container.getId())) {
+        if (containerIds.contains(container.getId())) {
```

Comment:
Sorry, values is a `Collection` which is what I was in fact referring to. Calling `map.values().contains(foo)` repeatedly as we add containers will not have `O(# existing containers) * O(# the added containers)` complexity, because `map.values()` does not need to be reconstructed for every value added, which is what I thought you were implying. The same backing collection gets used in the map impl, as new items are inserted so performance is the same as if we manage our own collections of values